tree/tree/src/TTree.cxx

```diff
@@ -4457,97 +4454,86 @@ Int_t TTree::Fill()
-               if (zipBytes != 0) {
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/zipBytes)/fEntries));
-               } else if (totBytes != 0) {
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/totBytes)/fEntries));
-               } else {
-                  TBufferFile b(TBuffer::kWrite, 10000);
-                  TTree::Class()->WriteBuffer(b, (TTree*) this);
-                  Long64_t total = b.Length();
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/total)/fEntries));
-               }
-            } else if(fAutoSave > 0) {
-               fAutoSave = fAutoFlush*(fAutoSave/fAutoFlush);
-            }
-            if (fAutoSave!=0 && fEntries >= fAutoSave) AutoSave();    // FlushBaskets not called in AutoSave
-            if (gDebug > 0) Info("TTree::Fill","First AutoFlush.  fAutoFlush = %lld, fAutoSave = %lld\n", fAutoFlush, fAutoSave);
-         }
-      } else if (fNClusterRange && fAutoFlush && ( (fEntries-fClusterRangeEnd[fNClusterRange-1]) % fAutoFlush == 0)  ) {
-         if (fAutoSave != 0 && fEntries%fAutoSave == 0) {
-            //We are at an AutoSave point. AutoSave flushes baskets and saves the Tree header
-            AutoSave("flushbaskets");
-            if (gDebug > 0) Info("TTree::Fill","AutoSave called at entry %lld, fZipBytes=%lld, fSavedBytes=%lld\n",fEntries,GetZipBytes(),fSavedBytes);
-         } else {
-            //We only FlushBaskets
-            FlushBaskets();
-            if (gDebug > 0) Info("TTree::Fill","FlushBasket called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",fEntries,GetZipBytes(),fFlushedBytes);
-         }
-         fFlushedBytes = GetZipBytes();
-      } else if (fNClusterRange == 0 && fEntries > 1 && fAutoFlush && fEntries%fAutoFlush == 0) {
-         if (fAutoSave != 0 && fEntries%fAutoSave == 0) {
-            //We are at an AutoSave point. AutoSave flushes baskets and saves the Tree header
-            AutoSave("flushbaskets");
-            if (gDebug > 0) Info("TTree::Fill","AutoSave called at entry %lld, fZipBytes=%lld, fSavedBytes=%lld\n",fEntries,GetZipBytes(),fSavedBytes);
-         } else {
-            //We only FlushBaskets
-            FlushBaskets();
-            if (gDebug > 0) Info("TTree::Fill","FlushBasket called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",fEntries,GetZipBytes(),fFlushedBytes);
+
+            // recompute auto save as number of entries if expressed in bytes
+            if (fAutoSave < 0)
+               fAutoSave = -fAutoSave/(zipBytes > 0 ? zipBytes : GetTotBytes());
+
+            // make fAutoSave a multiple of fAutoFlush
+            fAutoSave = TMath::Max(fAutoFlush, fAutoFlush * (fAutoSave/fAutoFlush));
+
+            if (gDebug > 0)
+               Info("TTree::Fill", "First AutoFlush. fAutoFlush = %lld, fAutoSave = %lld\n", fAutoFlush, fAutoSave);
          }
-         fFlushedBytes = GetZipBytes();
       }
    }
+
+   if (autoFlush) {
```

Comment:
Previously we had either `FlushBaskets()` or `AutoSave("flushbaskets")`. Now we just call them separately when auto saving. No real difference.

Reply:
So AutoSave no longer calls FlushBaskets? [And if it no longer does, is that really semantically correct since AutoSave can be called independently of this routine?]

Reply:
No, `AutoSave()` no longer calls `FlushBaskets()`. However, whenever it would call it, `autoFlush` is set to `true`, so it gets called beforehand and there is no real change in behavior. The only diference is that instead of calling `AutoSave("flushbaskets");` we call `FlushBaskets(); AutoSave();` instead. This is because for performance reasons (i.e. to shorten branches in the `if`s) I wanted to have only one place where these functions are called in TTree::Fill(), instead of calling them in several places.